rfcs/convergence/async-import-hidden-components.md

```diff
@@ -0,0 +1,88 @@
+
+## Summary
+
+When components such as `Tooltip` and `ContextualMenu` are not present in the DOM, they should be asynchronously imported to avoid bundle size related issues.
+
+## Background
+
+In the past we've avoided the use of async imports to prescribe split points, because the technology was new, and bundlers didn't have strong support for it. We built a proprietary webpack plugin for managing re-aliasing. This was done before webpack supported async import split points, so that we could ensure the right files were aliased, rather than asking each partner to set this up. But (I believe) the only partner to take advantage of this plugin was OOUI. (Was very proprietary, not well documented, etc.) But now with async imports, we don't need plugins. With the new components being worked on and tools being modernized, we should revisit this issue.
+
+If you don't know what they are; it is a way to asynchronously import source code. Bundlers use async imports as a cue for creating a split point in the bundle.
+
+Read more about how React itself helps make this simple for components: https://reactjs.org/docs/code-splitting.html
+
+## Problem statement
+
+We have several scenarios where async imports would be beneficial for minimizing bundle size impact on our partners apps, especially for initial page load.
+
+When a page loads, there is no need to load the code for `ContextualMenu` or positioning logic used by `Tooltip`/`Keytip`. That code should parse and initialize only when needed; when the user clicks a button to show a menu or hovers over an item that shows the `Tooltip`.
+
+Currently, we are avoiding using our own components in meaningful places, with (unnecessary) concern of bundle bloat due to this issue.
+
+Partners will likely do the same. (e.g. try to use `Tooltip`, find out they regressed 10k on initial paint impact and use an alternative solution.)
+
+## Detailed Design or Proposal
+
+Components such as `Tooltip` are split into 2 parts:
+
+1. A synchronous tiny component
+
+When a consumer imports `Tooltip`, they get the tiny part in their bundle which just registers handlers and logic to determine when the tooltip is visible.
+
+```jsx
+<>
+{state.children}
+{state.shouldRenderTooltip && (...)}
+</>
+```
+
+2. A asynchronous heavy component
+
+Once it becomes visible, we async import the actual `Tooltip` code, which in turn has graph edges to heavy things like `Popper` and styling.
+
+```jsx
+<Portal>
+  <slots.root {...slotProps.root}>
+    {state.pointing && <div ref={state.arrowRef} className={state.arrowClassName} />}
+    <slots.content {...slotProps.content} />
+  </slots.root>
+</Portal>
+```
+
+### Example Implementation
+
+```jsx
+const RenderTooltipContent = React.lazy(() => import('./renderTooltipContent'));
+
+export const renderTooltip = (state: TooltipState) => {
+  return (
+    <>
+      {state.children}
+      {state.shouldRenderTooltip && (
+        <React.Suspense fallback={null}>
```

Comment:
- What happens when this fails to load/loads after `.shouldRenderTooltip` becomes `true`? Customers will see nothing?
- SSR is not supported by `.Suspense` ➡ we can't rely on it currently

----

Please check for inspiration: https://github.com/theKashey/use-sidecar
I haven't checked how it works, but it's very close to our goal I guess

Reply:
If a javascript file fails to load on web app startup or on an async on demand bundle, this is essentially out of scope of fluent ui. Some webpack partners override the webpack async loading logic to have retries and fallbacks and ultimately a BSOD equivallent.